src/ergw_tei_mngr.erl

```diff
@@ -0,0 +1,168 @@
+%% modify it under the terms of the GNU General Public License
+%% as published by the Free Software Foundation; either version
+%% 2 of the License, or (at your option) any later version.
+
+%% simple implementation of 3GPP TS 29.598 Nudsf service
+%% - functional API according to TS
+%% - not optimized for anything
+
+-module(ergw_tei_mngr).
+
+-behavior(gen_statem).
+
+-compile({parse_transform, cut}).
+
+%% API
+-export([start_link/0, alloc_tei/1, release_tei/2]).
+
+%% gen_statem callbacks
+-export([callback_mode/0, init/1, handle_event/4, terminate/3, code_change/4]).
+
+-export([validate_option/1]).
+
+-include("include/ergw.hrl").
+
+-define(SERVER, ?MODULE).
+-define(SAVE_COUNTER, 10).
+-define(SAVE_TIMEOUT, 1000).
+-define(MAX_TRIES, 32).
+-define(POW(X, Y), trunc(math:pow(X, Y))).
+
+-record(data, {node_id, block_id, prefix, prefix_len, tid, save_cnt}).
+
+%%%=========================================================================
+%%%  API
+%%%=========================================================================
+
+start_link() ->
+    gen_statem:start_link({local, ?SERVER}, ?MODULE, [], []).
+
+%% with_keyfun/2
+with_keyfun(#request{gtp_port = Port}, Fun) ->
+    with_keyfun(Port, Fun);
+with_keyfun(#gtp_port{name = Name} = Port, Fun) ->
+    Fun(Name, gtp_context:port_teid_key(Port, _));
+with_keyfun(#pfcp_ctx{name = Name} = PCtx, Fun) ->
+    Fun(Name, ergw_pfcp:ctx_teid_key(PCtx, _)).
+
+%% alloc_tei/1
+alloc_tei(Port) ->
+    with_keyfun(Port, fun alloc_tei/2).
+
+%% alloc_tei/2
+alloc_tei(Name, KeyFun) ->
+    gen_server:call(?SERVER, {alloc_tei, Name, KeyFun}).
+
+%% release_tei/2
+release_tei(Port, TEI) ->
+    with_keyfun(Port, release_tei(_, TEI, _)).
+
+%% release_tei/3
+release_tei(Name, TEI, KeyFun) ->
+    gen_server:call(?SERVER, {release_tei, Name, TEI, KeyFun}).
+
+%%%===================================================================
+%%% Options Validation
+%%%===================================================================
+
+validate_option({Prefix, Len} = Value)
+  when is_integer(Prefix), is_integer(Len),
+       Len >= 0, Len =< 8, Prefix >= 0 ->
+    case ?POW(2, Len) of
+	X when X > Prefix -> Value;
+	_ ->
+	    throw({error, {options, {teid, Value}}})
+    end;
+validate_option(Value) ->
+    throw({error, {options, {teid, Value}}}).
+
+%%%===================================================================
+%%% gen_statem callbacks
+%%%===================================================================
+
+callback_mode() -> handle_event_function.
+
+init([]) ->
+    {ok, {Prefix, PrefixLen}} = application:get_env(ergw, teid),
+    TID = ets:new(?SERVER, [ordered_set, {keypos, 1}]),
+    NodeId = node(),
+    BlockId = iolist_to_binary(["teid-prefix-", integer_to_list(Prefix)]),
+
+    Data = #data{node_id = NodeId,
+		  block_id = BlockId,
+		  prefix = Prefix,
+		  prefix_len = PrefixLen,
+		  tid = TID,
+		  save_cnt = ?SAVE_COUNTER},
```

Comment:
node_id, block_id and save_cnt are not really use in this version. Even when we later use them, adding them in this version is not needed.